### PR TITLE
[AArch64][GlobalISel] Support more types for TRUNC

### PR DESCRIPTION
**Pull request by chuongg3**
G_TRUNC will get lowered into trunc(merge(trunc(unmerge), trunc(unmerge))) if the source is larger than 128 bits or the truncation is more than half of the current bit size.

Now mirrors ZEXT/SEXT code more closely for vector types.

### CODE REVIEW

File: llvm/test/CodeGen/AArch64/xtn.ll [hunk 1]
```diff
@@ -0,0 +1,473 @@
...
+; CHECK-GI:       // %bb.0: // %entry
+; CHECK-GI-NEXT:    fmov x0, d0
+; CHECK-GI-NEXT:    fmov x1, d1
+; CHECK-GI-NEXT:    fmov x2, d2
+; CHECK-GI-NEXT:    // kill: def $w0 killed $w0 killed $x0
+; CHECK-GI-NEXT:    // kill: def $w1 killed $w1 killed $x1
+; CHECK-GI-NEXT:    // kill: def $w2 killed $w2 killed $x2
+; CHECK-GI-NEXT:    ret
+entry:
+  %arg1 = trunc <3 x i64> %a to <3 x i8>
```
Comment (tschuett) [inline llvm/test/CodeGen/AArch64/xtn.ll:295]: I like this case.

File: llvm/lib/Target/AArch64/GISel/AArch64LegalizerInfo.cpp [hunk 2]
```diff
@@ -542,14 +542,22 @@ AArch64LegalizerInfo::AArch64LegalizerInfo(const AArch64Subtarget &ST)
       });
 
   getActionDefinitionsBuilder(G_TRUNC)
+      .legalFor({{v2s32, v2s64}, {v4s16, v4s32}, {v8s8, v8s16}})
```
Comment (tschuett) [inline llvm/lib/Target/AArch64/GISel/AArch64LegalizerInfo.cpp:545]: Any plans for a MIR test as you changed the legalizer?

Comment (chuongg3) [reply llvm/lib/Target/AArch64/GISel/AArch64LegalizerInfo.cpp:545]: I have added some MIR tests in the follow up patch, is this what you were looking for?

File: llvm/test/CodeGen/AArch64/GlobalISel/legalize-xtn.mir [hunk 3]
```diff
@@ -0,0 +1,531 @@
...
+    ; CHECK-LABEL: name: xtn_v4i32_v4i8
+    ; CHECK: liveins: $q0
+    ; CHECK-NEXT: {{  $}}
+    ; CHECK-NEXT: [[COPY:%[0-9]+]]:_(<4 x s32>) = COPY $q0
+    ; CHECK-NEXT: [[TRUNC:%[0-9]+]]:_(<4 x s16>) = G_TRUNC [[COPY]](<4 x s32>)
+    ; CHECK-NEXT: $d0 = COPY [[TRUNC]](<4 x s16>)
+    ; CHECK-NEXT: RET_ReallyLR implicit $d0
+    %0:_(<4 x s32>) = COPY $q0
+    %1:_(<4 x s8>) = G_TRUNC %0(<4 x s32>)
+    %2:_(<4 x s16>) = G_ANYEXT %1(<4 x s8>)
```
Comment (tschuett) [inline llvm/test/CodeGen/AArch64/GlobalISel/legalize-xtn.mir:361]: The G_ANYEXT disappeared?

File: llvm/test/CodeGen/AArch64/zext.ll [hunk 4]
```diff
@@ -1089,67 +1089,35 @@ entry:
...
-; CHECK-GI-NEXT:    mov v5.s[3], w7
-; CHECK-GI-NEXT:    mov v0.s[3], v1.s[0]
-; CHECK-GI-NEXT:    mov v2.s[3], v3.s[0]
-; CHECK-GI-NEXT:    ldr q3, [x8, :lo12:.LCPI52_0]
-; CHECK-GI-NEXT:    uzp1 v1.8h, v4.8h, v5.8h
-; CHECK-GI-NEXT:    uzp1 v2.8h, v0.8h, v2.8h
-; CHECK-GI-NEXT:    and v0.16b, v1.16b, v3.16b
-; CHECK-GI-NEXT:    and v1.16b, v2.16b, v3.16b
-; CHECK-GI-NEXT:    ret
+; CHECK-LABEL: zext_v16i10_v16i16:
```
Comment (aemerson) [inline llvm/test/CodeGen/AArch64/zext.ll:1092]: Why are we now falling back for this test?

Comment (chuongg3) [reply llvm/test/CodeGen/AArch64/zext.ll:1092]: Due to the nature in which the parameters are loaded for this function, it gets created as a `v16i32` buildvector.
The parameter then gets truncated to `v16i10`, before being extended to `v16i16`.

**Review by davemgreen — APPROVED** [review]
This looks OK to me, from comparing it to the zext/sext version. LGTM if there are no other comments, thanks.

**Pull request merged**